### PR TITLE
Describe query option to infer labels.

### PR DESCRIPTION
**Pull request by arnefischereit**
New query option (aka pre-parser option) for inferring labels.

Also see https://github.com/neo4j/docs-operations/pull/1620

### CODE REVIEW

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc [hunk 1]
```diff
@@ -303,3 +303,27 @@ CYPHER replan=force EXPLAIN MATCH ...
...
+This option allows to control to what extent it should do that:
+
+[options="header",cols="2m,3a"]
+|===
+| Option | Description
+
+| inferSchemaParts=off
+| No predicates are inferred.
+
+| inferSchemaParts=most_selecive_label
```
Comment (WilcoNeo) [inline modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]:
```suggestion
| inferSchemaParts=most_selective_label
```

**Review by WilcoNeo — APPROVED** [review]

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc [hunk 2]
```diff
@@ -303,3 +303,27 @@ CYPHER replan=force EXPLAIN MATCH ...
...
+| Relationship types are used to infer labels on the relationships' end nodes and the label corresponding to the least number of nodes is used to estimate rows.
```
Comment (blouerat) [inline modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]:
```suggestion
| Relationship types are used to infer labels on the relationships' end nodes and the label corresponding to the smallest number of nodes is used to estimate rows.
```

**Review by blouerat — APPROVED** [review]

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc [hunk 3]
```diff
@@ -303,3 +303,28 @@ CYPHER replan=force EXPLAIN MATCH ...
 
 During times of known high load, `replan=skip` can be useful to not introduce unwanted latency spikes.
 
+[[cypher-infer-schema-parts]]
+=== Cypher infer schema parts
+
+For some queries, the planner can infer predicates such as labels or types from the graph structure that can improve estimating the number of rows that each operator produces.
+(See xref:planning-and-tuning/execution-plans.adoc#runtimes-reading-execution-plans[Cypher Manual -> Execution plans and query tuning -> Understanding execution plans])
+This option allows to control to what extent it should do that:
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]:
```suggestion
For some queries, the planner can infer predicates such as labels or types from the graph structure, thereby enhancing its ability to estimate the number of rows each operator will produce.
(See xref:planning-and-tuning/execution-plans.adoc#runtimes-reading-execution-plans[Understanding execution plans - Reading execution plans] for more information about the role of operators in query execution plans.)
The option `inferSchemaParts` controls the extent to which the planner should infer predicates.
```

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc [hunk 4]
```diff
@@ -303,3 +303,28 @@ CYPHER replan=force EXPLAIN MATCH ...
 
 During times of known high load, `replan=skip` can be useful to not introduce unwanted latency spikes.
 
+[[cypher-infer-schema-parts]]
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning.adoc:307]:
```suggestion
[role=label--new-5.21]
[[cypher-infer-schema-parts]]
```

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc [hunk 5]
```diff
@@ -303,3 +303,28 @@ CYPHER replan=force EXPLAIN MATCH ...
...
+|===
+| Option | Description
+
+| inferSchemaParts=off
+| No predicates are inferred.
+
+| inferSchemaParts=most_selective_label
+| Relationship types are used to infer labels on the relationships' end nodes.
+The label corresponding to the smallest number of nodes is used to estimate rows.
+Not inferring more labels yields better results for nodes with several dependent labels, such as every `:Actor` being a `:Person`.
```
Comment (JPryce-Aklundh) [inline modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]:
```suggestion
| Relationship types are used to infer labels on the end nodes of relationships.
The label corresponding to the smallest number of nodes is used to estimate rows.
Avoiding the inference of multiple labels improves accuracy for nodes with several dependent labels, such as every `:Actor` being a `:Person`.
```

**Review by JPryce-Aklundh — COMMENTED** [review]
Thanks @arnefischereit - just a few editorial suggestions from me. Also, we should probably add this to the Additions and deprecations page.

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc (hunk 3 above)
Comment (arnefischereit) [reply modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]: I think your version sounds a lot better. But I especially wanted to reference our only documentation on estimated rows.

What do you think about
```suggestion
For some queries, the planner can infer predicates such as labels or types from the graph structure, thereby enhancing its ability to estimate the number of rows each operator will produce.
(See xref:planning-and-tuning/execution-plans.adoc#runtimes-reading-execution-plans[Understanding execution plans - Reading execution plans] for more information about the role of operators and estimated row counts in query execution plans.)
The option `inferSchemaParts` controls the extent to which the planner should infer predicates.
```

Comment (JPryce-Aklundh) [reply modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]: Sounds good 👍

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

File: modules/ROOT/pages/planning-and-tuning/query-tuning.adoc [hunk 6]
```diff
@@ -303,3 +303,29 @@ CYPHER replan=force EXPLAIN MATCH ...
...
+
+[options="header",cols="2m,3a"]
+|===
+| Option | Description
+
+| inferSchemaParts=off
+| No predicates are inferred.
+
+| inferSchemaParts=most_selective_label
+| Relationship types are used to infer labels on the end nodes of relationships.
```
Comment (sherfert) [inline modules/ROOT/pages/planning-and-tuning/query-tuning.adoc]: I would consider not writing "end nodes", as that could be confused with just the nodes the relationships are pointing _to_.

"boundary" nodes? "connected" nodes?

**Comment by neo-technology-commit-status-publisher** [comment]
This PR includes documentation updates
View the updated docs at https://neo4j-docs-cypher-958.surge.sh

Updated pages:
- [Deprecations, additions, and compatibility](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/deprecations-additions-removals-compatibility)
- [Query tuning](https://neo4j-docs-cypher-958.surge.sh/cypher-manual/5/planning-and-tuning/query-tuning)

**Comment by neo-technology-commit-status-publisher** [comment]
Thanks for the documentation updates.

The preview documentation has now been torn down - reopening this PR will republish it.

**Pull request merged**